cpp/bench/ann/src/raft/raft_cagra_hnswlib_wrapper.h

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/device_resources.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/operators.hpp>
+#include <raft/distance/distance_types.hpp>
+#include <raft/linalg/unary_op.cuh>
+#include <raft/neighbors/cagra.cuh>
+#include <raft/neighbors/cagra_serialize.cuh>
+#include <raft/neighbors/cagra_types.hpp>
+#include <raft/util/cudart_utils.hpp>
+#include <rmm/device_uvector.hpp>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+
+#include "../common/ann_types.hpp"
+#include "../common/thread_pool.hpp"
+#include "raft_ann_bench_utils.h"
+#include <raft/util/cudart_utils.hpp>
+
+#include <hnswlib.h>
+
+namespace raft::bench::ann {
+
+template <typename T>
+struct hnsw_dist_t {
+  using type = void;
+};
+
+template <>
+struct hnsw_dist_t<float> {
+  using type = float;
+};
+
+template <>
+struct hnsw_dist_t<uint8_t> {
+  using type = int;
+};
+
+template <>
+struct hnsw_dist_t<int8_t> {
+  using type = int;
+};
+
+template <typename T, typename IdxT>
+class RaftCagraHnswlib : public ANN<T> {
+ public:
+  using typename ANN<T>::AnnSearchParam;
+
+  struct SearchParam : public AnnSearchParam {
+    int ef;
+    int num_threads = 1;
+  };
+
+  using BuildParam = raft::neighbors::cagra::index_params;
+
+  RaftCagraHnswlib(Metric metric, int dim, const BuildParam& param, int concurrent_searches = 1)
+    : ANN<T>(metric, dim), index_params_(param), dimension_(dim), handle_(cudaStreamPerThread)
+  {
+    index_params_.metric = parse_metric_type(metric);
+    RAFT_CUDA_TRY(cudaGetDevice(&device_));
+  }
+
+  ~RaftCagraHnswlib() noexcept {}
+
+  void build(const T* dataset, size_t nrow, cudaStream_t stream) final;
+
+  void set_search_param(const AnnSearchParam& param) override;
+
+  // TODO: if the number of results is less than k, the remaining elements of 'neighbors'
+  // will be filled with (size_t)-1
+  void search(const T* queries,
```

Comment:
We should be able to reuse all this from hnswlib's wrapper, right? We should consolidate this stuff because having duplicated code all over the place is going to make this hard to maintain.